consensus/src/dag/mod.rs

```diff
@@ -0,0 +1,641 @@
+                DagResult::empty()
+            }
+            NewForker(alert) => {
+                warn!(target: LOG_TARGET, "New forker detected.");
+                trace!(target: LOG_TARGET, "Created alert: {:?}.", alert);
+                DagResult::alert(*alert)
+            }
+        }
+    }
+
+    fn handle_result(&mut self, result: &DagResult<H, D, MK>) {
+        // just clean the validator cache of units that we are returning
+        for unit in &result.units {
+            self.validator.finished_processing(&unit.hash());
+        }
+    }
+
+    /// Add a unit to the Dag.
+    pub fn add_unit<U: WrappedUnit<H, Wrapped = SignedUnit<H, D, MK>>>(
+        &mut self,
+        unit: UncheckedSignedUnit<H, D, MK::Signature>,
+        store: &UnitStore<U>,
+    ) -> DagResult<H, D, MK> {
+        let result = match self.validator.validate(unit, store) {
+            Ok(unit) => self.reconstruction.add_unit(unit).into(),
+            Err(e) => Self::handle_validation_error(e),
+        };
+        self.handle_result(&result);
+        result
+    }
+
+    /// Add parents of a unit to the Dag.
+    pub fn add_parents<U: WrappedUnit<H, Wrapped = SignedUnit<H, D, MK>>>(
+        &mut self,
+        unit_hash: H::Hash,
+        parents: Vec<UncheckedSignedUnit<H, D, MK::Signature>>,
+        store: &UnitStore<U>,
+    ) -> DagResult<H, D, MK> {
+        use ValidationError::*;
+        let mut result = DagResult::empty();
+        let mut parent_hashes = HashMap::new();
+        for unit in parents {
+            let unit = match self.validator.validate(unit, store) {
+                Ok(unit) => {
+                    result.accumulate(self.reconstruction.add_unit(unit.clone()).into());
+                    unit
+                }
+                Err(Invalid(e)) => {
+                    warn!(target: LOG_TARGET, "Received parent failing validation: {}", e);
+                    return result;
+                }
+                Err(Duplicate(unit)) => {
+                    trace!(target: LOG_TARGET, "Received parent with hash {:?} again.", unit.hash());
+                    unit
+                }
+                Err(Uncommitted(unit)) => {
+                    debug!(target: LOG_TARGET, "Received uncommitted parent {:?}, we should get the commitment soon.", unit.hash());
+                    unit
+                }
+                Err(NewForker(alert)) => {
+                    warn!(target: LOG_TARGET, "New forker detected.");
+                    trace!(target: LOG_TARGET, "Created alert: {:?}.", alert);
+                    result.add_alert(*alert);
+                    // technically this was a correct unit, so we could have passed it on,
+                    // but this will happen at most once and we will receive the parent
+                    // response again, so we just discard it now
+                    return result;
```

Comment:
Why do we return already? I felt unsure about it, because some prefix of `parents` which passed validation will be included in the `result`, but some suffix of `parents` which would have passed validation as well won't be included. I would presume that the ordering in `parents` is arbitrary, and what worries me that `result` depends on this ordering (i.e. which of the parents caused a fork).